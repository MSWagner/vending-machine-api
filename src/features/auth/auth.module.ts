import { Module } from "@nestjs/common";

import { AuthService } from "./services/auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";
import { LoginController } from "./controller/login/login.controller";
import { RefreshTokenController } from "./controller/refresh-token/refresh-token.controller";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    controllers: [LoginController, RefreshTokenController],
    exports: [AuthService]
})
export class AuthModule {}
