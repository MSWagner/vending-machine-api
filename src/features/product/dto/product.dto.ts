import { IsDivisibleBy, IsInt, IsString, Min, MinLength } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";

export class ProductUpsertDto {
    @IsString()
    @MinLength(3)
    @ApiProperty({ minLength: 3 })
    readonly productName: string;

    @IsInt()
    @Min(5)
    @IsDivisibleBy(5)
    @ApiProperty({ minimum: 0 })
    readonly cost: number;

    @IsInt()
    @Min(0)
    @ApiProperty({ minimum: 0 })
    readonly amountAvailable: number;
}

export class ProductResponseDto extends ProductUpsertDto {
    @ApiProperty({ format: "uuid" })
    readonly uid: string;

    @ApiProperty({ type: "date" })
    readonly createdAt: Date;

    @ApiProperty({ type: "date" })
    readonly updatedAt: Date;
}
