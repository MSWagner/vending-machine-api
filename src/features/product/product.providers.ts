import { Connection, Repository } from "typeorm";
import { Product } from "../../entities/Product.entity";

import CONFIG from "../../config";

export const productProviders = [
    {
        provide: CONFIG.database.defaultProductRepoName,
        useFactory: (connection: Connection): Repository<Product> => connection.getRepository(Product),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
