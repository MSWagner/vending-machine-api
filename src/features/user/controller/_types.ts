import { ApiProperty } from "@nestjs/swagger";

export class UserPostResponse {
    @ApiProperty()
    readonly tokenType: string;

    @ApiProperty()
    readonly expiresIn: number;

    @ApiProperty({ format: "uuid" })
    refreshToken: string;

    @ApiProperty({ format: "uuid" })
    accessToken: string;
}

export class UserGetResponse {
    @ApiProperty()
    readonly username: string;

    @ApiProperty()
    readonly role: string;

    @ApiProperty()
    readonly deposit?: number;
}

export class SuccessResponse {
    @ApiProperty()
    readonly success: boolean;
}
