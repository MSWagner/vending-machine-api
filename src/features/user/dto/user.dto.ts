import { IsEnum, IsString, MinLength } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { Role } from "../../../entities/User.entity";

export class UserDto {
    @IsString()
    @MinLength(3)
    @ApiProperty({ minLength: 3 })
    readonly username: string;

    @IsString()
    @MinLength(6)
    @ApiProperty({ minLength: 6 })
    readonly password: string;

    @IsString()
    @MinLength(3)
    @IsEnum(Role)
    @ApiProperty({ minLength: 6 })
    readonly role: Role;
}
