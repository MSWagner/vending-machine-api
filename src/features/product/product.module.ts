import { Module } from "@nestjs/common";
import { ProductService } from "./services/product.service";
import { productProviders } from "./product.providers";

import { DatabaseModule } from "../../shared/database/database.module";
import { UserModule } from "../user/user.module";
import { AuthModule } from "../auth/auth.module";
import { ProductController } from "./controller/product.controller";

@Module({
    imports: [DatabaseModule, UserModule, AuthModule],
    providers: [...productProviders, ProductService],
    controllers: [ProductController],
    exports: [DatabaseModule, ProductService, ...productProviders]
})
export class ProductModule {}
