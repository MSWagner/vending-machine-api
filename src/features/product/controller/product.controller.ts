import _ from "lodash";
import {
    Controller,
    Request,
    Post,
    Body,
    Param,
    Delete,
    Put,
    UseGuards,
    Get,
    HttpException,
    HttpStatus
} from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { ApiCreatedResponse, ApiTags } from "@nestjs/swagger";

import { Roles } from "../../auth/guards/role.decorator";
import { IPublicProductData, Product, Role } from "../../../entities/Product.entity";
import { ProductService } from "../services/product.service";
import { RoleGuard } from "../../auth/guards/role.guard";
import { ProductUpsertDto, ProductResponseDto } from "../dto/product.dto";
import { SuccessResponse } from "./_types";

@ApiTags("products")
@Controller({ path: "products", version: "1" })
export class ProductController {
    constructor(private readonly productService: ProductService) {}

    @Post()
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Seller)
    @ApiCreatedResponse({ description: "The product has been successfully created.", type: ProductResponseDto })
    async create(@Request() req, @Body() productDto: ProductUpsertDto): Promise<IPublicProductData> {
        return (await this.productService.create(req.user, productDto)).publicData;
    }

    @Put(":uid")
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Seller)
    @ApiCreatedResponse({ description: "The product has been successfully updated.", type: ProductResponseDto })
    async update(
        @Request() req,
        @Param("uid") uid: string,
        @Body() productDto: ProductUpsertDto
    ): Promise<IPublicProductData> {
        const product = await this.productService.findOne(uid);

        this.checkValidProductOwner(product, req.user.uid);

        return (await this.productService.update(product, req.user, productDto)).publicData;
    }

    @Get()
    @ApiCreatedResponse({ type: ProductResponseDto, isArray: true })
    async findAll(): Promise<IPublicProductData[]> {
        return (await this.productService.findAll()).map((product) => product.publicData);
    }

    @Get(":uid")
    @ApiCreatedResponse({ type: ProductResponseDto })
    async findOne(@Param("uid") uid: string): Promise<IPublicProductData> {
        const product = await this.productService.findOne(uid);

        if (_.isNil(product)) {
            throw new HttpException("Product could not be found.", HttpStatus.NOT_FOUND);
        } else {
            return product.publicData;
        }
    }

    @Delete(":uid")
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Seller)
    @ApiCreatedResponse({ description: "The product has been successfully deleted.", type: SuccessResponse })
    async remove(@Request() req, @Param("uid") uid: string): Promise<SuccessResponse> {
        const product = await this.productService.findOne(uid);

        this.checkValidProductOwner(product, req.user.uid);

        await this.productService.delete(product);

        return { success: true };
    }

    private checkValidProductOwner(product: Product, userUid: string): void {
        if (_.isNil(product)) {
            throw new HttpException("Product could not be found.", HttpStatus.NOT_FOUND);
        }

        if (product.seller.uid !== userUid) {
            throw new HttpException("You are not the owner of this product.", HttpStatus.UNAUTHORIZED);
        }
    }
}
