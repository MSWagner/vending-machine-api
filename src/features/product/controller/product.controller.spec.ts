import request from "supertest";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";

import { setupApp } from "../../../appSetup";

import { TestModule } from "../../../shared/test/test.module";
import { TestService } from "../../../shared/test/test.service";
import { AuthModule } from "../../../features/auth/auth.module";
import { UserModule } from "../../../features/user/user.module";
import { ProductController } from "./product.controller";
import { ProductUpsertDto } from "../dto/product.dto";
import { Product } from "../../../entities/Product.entity";
import { ProductService } from "../services/product.service";
import { productProviders } from "../product.providers";

import * as fixtures from "../../../shared/test/fixtures";

describe("ProductController", () => {
    let app: INestApplication;
    let testService: TestService;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            controllers: [ProductController],
            providers: [...productProviders, ProductService],
            imports: [TestModule, AuthModule, UserModule]
        }).compile();

        testService = module.get<TestService>(TestService);

        app = module.createNestApplication();
        await setupApp(app);
        await app.init();
    });

    beforeEach(async () => {
        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    //// POST /products

    it("/POST products - should create a new product for a authenticated seller", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(201);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NewProduct1CreatedResponse");
    });

    it("/POST products - should throw an error for unauthorized requests", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer()).post("/api/v1/products").send(newProduct).expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("UnauthorizedPostError");
    });

    it("/POST products - should throw an error for authentictated buyer requests", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(newProduct)
            .expect(403);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("BuyerPostError");
    });

    it("/POST products - should throw an error for missing productName", async () => {
        const newProduct = {
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("ProductNameMissingPostError");
    });

    it("/POST products - should throw an error for missing cost", async () => {
        const newProduct = {
            productName: "newProduct1",
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("CostMissingPostError");
    });

    it("/POST products - should throw an error for missing amountAvailable", async () => {
        const newProduct = {
            productName: "newProduct1",
            cost: 25
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("AmountAvailableMissingPostError");
    });

    it("/POST products - should throw an error for cost is not be in multiples of 5", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 7,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("CostNotMultipleOf5PostError");
    });

    it("/POST products - should throw an error for cost is to low", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 0,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("CostToLowPostError");
    });

    it("/POST products - should throw an error for amountAvailable is to low", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 5,
            amountAvailable: -1
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("amountAvailableLowPostError");
    });

    //// GET /products

    it("/GET products - should return all products for unauthenticated users", async () => {
        const response = await request(app.getHttpServer()).get("/api/v1/products").expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("AllProductsResponse");
    });

    it("/GET products - should return all products for buyer users", async () => {
        const response = await request(app.getHttpServer())
            .get("/api/v1/products")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("AllProductsResponse");
    });

    it("/GET products - should return all products for seller users", async () => {
        const response = await request(app.getHttpServer())
            .get("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("AllProductsResponse");
    });

    it("/GET products - should return an empty array if no products available", async () => {
        await Product.clear();
        const allProductCount = await Product.count();
        expect(allProductCount).toBe(0);

        const response = await request(app.getHttpServer())
            .get("/api/v1/products")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("EmptyProductsResponse");
    });

    //// GET /products/:uid

    it("/GET products/:uid - should return products1 for unauthenticated users", async () => {
        const response = await request(app.getHttpServer())
            .get("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("Product1Response");
    });

    it("/GET products/:uid - should return products1 for buyer users", async () => {
        const response = await request(app.getHttpServer())
            .get("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("Product1Response");
    });

    it("/GET products/:uid - should return products1 for seller users", async () => {
        const response = await request(app.getHttpServer())
            .get("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("Product1Response");
    });

    it("/GET products/:uid - should throw an 404 error for not found product", async () => {
        await Product.clear();
        const allProductCount = await Product.count();
        expect(allProductCount).toBe(0);

        const response = await request(app.getHttpServer())
            .get("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .expect(404);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("ErrorProductNotFoundResponse");
    });

    //// PUT /products/uid

    it("/PUT products/uid - should update a product1 from user3 (seller)", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1Name",
            cost: 25,
            amountAvailable: 9
        };

        const product1Before = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1Before).not.toBeNull();
        expect(product1Before.productName).not.toBe(newProduct.productName);
        expect(product1Before.cost).not.toBe(newProduct.cost);
        expect(product1Before.amountAvailable).not.toBe(newProduct.amountAvailable);

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(200);

        const product1After = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1After).not.toBeNull();
        expect(product1After.productName).toBe(newProduct.productName);
        expect(product1After.cost).toBe(newProduct.cost);
        expect(product1After.amountAvailable).toBe(newProduct.amountAvailable);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NewProduct1UpdatedResponse");
    });

    it("/PUT products/uid  - should throw an error for unauthorized requests", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .send(newProduct)
            .expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("UnauthorizedPutError");
    });

    it("/PUT products/uid  - should throw an error for authentictated buyer requests", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(newProduct)
            .expect(403);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("BuyerPutError");
    });

    it("/PUT products/uid - should throw an error for missing productName", async () => {
        const newProduct = {
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("ProductNameMissingPutError");
    });

    it("/PUT products/uid - should throw an error for missing cost", async () => {
        const newProduct = {
            productName: "newProduct1",
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("CostMissingPutError");
    });

    it("/PUT products/uid - should throw an error for missing amountAvailable", async () => {
        const newProduct = {
            productName: "newProduct1",
            cost: 25
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("AmountAvailableMissingPutError");
    });

    it("/PUT products/uid - should throw an error for cost is not be in multiples of 5", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 7,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("CostNotMultipleOf5PutError");
    });

    it("/PUT products/uid - should throw an error for cost is to low", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 0,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("CostToLowPutError");
    });

    it("/PUT products/uid - should throw an error for amountAvailable is to low", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 5,
            amountAvailable: -1
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(400);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("amountAvailableLowPutError");
    });

    it("/PUT products/uid  - should throw an error for a not owner seller requests", async () => {
        const newProduct: ProductUpsertDto = {
            productName: "newProduct1",
            cost: 25,
            amountAvailable: 9
        };

        const response = await request(app.getHttpServer())
            .put("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser4Seller.token, { type: "bearer" })
            .send(newProduct)
            .expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotOwnerPutError");
    });

    //// DELETE /products/uid

    it("/DELETE products/uid - should successful delete product1", async () => {
        const response = await request(app.getHttpServer())
            .delete("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .expect(200);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("DeleteProduct1");
    });

    it("/DELETE products/uid - should throw an error for no access", async () => {
        const response = await request(app.getHttpServer())
            .delete("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NoAccessError");
    });

    it("/DELETE products/uid - should throw an not owner request", async () => {
        const response = await request(app.getHttpServer())
            .delete("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser4Seller.token, { type: "bearer" })
            .expect(401);
        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotOwnerDeleteError");
    });

    it("/DELETE user - should throw an already deleted product", async () => {
        await Product.delete({ uid: fixtures.product1User3Seller.uid });

        const response = await request(app.getHttpServer())
            .delete("/api/v1/products/" + fixtures.product1User3Seller.uid)
            .auth(fixtures.accessTokenUser4Seller.token, { type: "bearer" })
            .expect(404);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("AlreadyDeletedError");
    });
});
