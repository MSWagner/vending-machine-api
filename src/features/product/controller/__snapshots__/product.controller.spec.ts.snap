// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductController /DELETE products/uid - should successful delete product1: DeleteProduct1 1`] = `
{
  "success": true,
}
`;

exports[`ProductController /DELETE products/uid - should throw an error for no access: NoAccessError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`ProductController /DELETE products/uid - should throw an not owner request: NotOwnerDeleteError 1`] = `
{
  "message": "You are not the owner of this product.",
  "statusCode": 401,
}
`;

exports[`ProductController /DELETE user - should throw an already deleted product: AlreadyDeletedError 1`] = `
{
  "message": "Product could not be found.",
  "statusCode": 404,
}
`;

exports[`ProductController /GET products - should return all products for buyer users: AllProductsResponse 1`] = `
[
  {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  {
    "amountAvailable": 1,
    "cost": 10,
    "createdAt": "createdAt",
    "productName": "product2",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1Seller4",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
]
`;

exports[`ProductController /GET products - should return all products for seller users: AllProductsResponse 1`] = `
[
  {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  {
    "amountAvailable": 1,
    "cost": 10,
    "createdAt": "createdAt",
    "productName": "product2",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1Seller4",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
]
`;

exports[`ProductController /GET products - should return all products for unauthenticated users: AllProductsResponse 1`] = `
[
  {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  {
    "amountAvailable": 1,
    "cost": 10,
    "createdAt": "createdAt",
    "productName": "product2",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1Seller4",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
]
`;

exports[`ProductController /GET products - should return an empty array if no products available: EmptyProductsResponse 1`] = `[]`;

exports[`ProductController /GET products/:uid - should return products1 for buyer users: Product1Response 1`] = `
{
  "amountAvailable": 10,
  "cost": 100,
  "createdAt": "createdAt",
  "productName": "product1",
  "uid": "uid",
  "updatedAt": "updatedAt",
}
`;

exports[`ProductController /GET products/:uid - should return products1 for seller users: Product1Response 1`] = `
{
  "amountAvailable": 10,
  "cost": 100,
  "createdAt": "createdAt",
  "productName": "product1",
  "uid": "uid",
  "updatedAt": "updatedAt",
}
`;

exports[`ProductController /GET products/:uid - should return products1 for unauthenticated users: Product1Response 1`] = `
{
  "amountAvailable": 10,
  "cost": 100,
  "createdAt": "createdAt",
  "productName": "product1",
  "uid": "uid",
  "updatedAt": "updatedAt",
}
`;

exports[`ProductController /GET products/:uid - should throw an 404 error for not found product: ErrorProductNotFoundResponse 1`] = `
{
  "message": "Product could not be found.",
  "statusCode": 404,
}
`;

exports[`ProductController /POST products - should create a new product for a authenticated seller: NewProduct1CreatedResponse 1`] = `
{
  "amountAvailable": 9,
  "cost": 25,
  "createdAt": "createdAt",
  "productName": "newProduct1",
  "uid": "uid",
  "updatedAt": "updatedAt",
}
`;

exports[`ProductController /POST products - should throw an error for amountAvailable is to low: amountAvailableLowPostError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "amountAvailable must not be less than 0",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /POST products - should throw an error for authentictated buyer requests: BuyerPostError 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`ProductController /POST products - should throw an error for cost is not be in multiples of 5: CostNotMultipleOf5PostError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "cost must be divisible by 5",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /POST products - should throw an error for cost is to low: CostToLowPostError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "cost must not be less than 5",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /POST products - should throw an error for missing amountAvailable: AmountAvailableMissingPostError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "amountAvailable must not be less than 0",
    "amountAvailable must be an integer number",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /POST products - should throw an error for missing cost: CostMissingPostError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "cost must be divisible by 5",
    "cost must not be less than 5",
    "cost must be an integer number",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /POST products - should throw an error for missing productName: ProductNameMissingPostError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "productName must be longer than or equal to 3 characters",
    "productName must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /POST products - should throw an error for unauthorized requests: UnauthorizedPostError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`ProductController /PUT products/uid  - should throw an error for a not owner seller requests: NotOwnerPutError 1`] = `
{
  "message": "You are not the owner of this product.",
  "statusCode": 401,
}
`;

exports[`ProductController /PUT products/uid  - should throw an error for authentictated buyer requests: BuyerPutError 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`ProductController /PUT products/uid  - should throw an error for unauthorized requests: UnauthorizedPutError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`ProductController /PUT products/uid - should throw an error for amountAvailable is to low: amountAvailableLowPutError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "amountAvailable must not be less than 0",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /PUT products/uid - should throw an error for cost is not be in multiples of 5: CostNotMultipleOf5PutError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "cost must be divisible by 5",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /PUT products/uid - should throw an error for cost is to low: CostToLowPutError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "cost must not be less than 5",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /PUT products/uid - should throw an error for missing amountAvailable: AmountAvailableMissingPutError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "amountAvailable must not be less than 0",
    "amountAvailable must be an integer number",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /PUT products/uid - should throw an error for missing cost: CostMissingPutError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "cost must be divisible by 5",
    "cost must not be less than 5",
    "cost must be an integer number",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /PUT products/uid - should throw an error for missing productName: ProductNameMissingPutError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "productName must be longer than or equal to 3 characters",
    "productName must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`ProductController /PUT products/uid - should update a product1 from user3 (seller): NewProduct1UpdatedResponse 1`] = `
{
  "amountAvailable": 9,
  "cost": 25,
  "createdAt": "createdAt",
  "productName": "newProduct1Name",
  "uid": "uid",
  "updatedAt": "updatedAt",
}
`;
